crates/rpc/src/dto/receipt.rs

```diff
@@ -33,11 +38,43 @@ pub enum TxnFinalityStatus {
     AcceptedOnL1,
 }
 
+pub struct TxnReceiptWithBlockInfo<'a> {
+    pub block_hash: Option<&'a BlockHash>,
+    pub block_number: Option<BlockNumber>,
+    pub receipt: &'a Receipt,
+    pub transaction: &'a Transaction,
+    pub events: &'a [Event],
+    pub finality: TxnFinalityStatus,
+}
+
+pub struct TxnReceipt<'a> {
+    pub receipt: &'a Receipt,
+    pub transaction: &'a Transaction,
+    pub events: &'a [Event],
+    pub finality: TxnFinalityStatus,
+}
+
+pub struct InvokeTxnReceipt<'a>(pub &'a TxnReceipt<'a>);
+pub struct L1HandlerTxnReceipt<'a>(pub &'a TxnReceipt<'a>);
+pub struct DeclareTxnReceipt<'a>(pub &'a TxnReceipt<'a>);
+pub struct DeployTxnReceipt<'a>(pub &'a TxnReceipt<'a>);
+pub struct DeployAccountTxnReceipt<'a>(pub &'a TxnReceipt<'a>);
```

Comment:
Its possible to have stricter type bounds here, but it would require multiple additional helper enums and structs. I instead elected for checks within serialzation.